File: q0137/q0137.cpp
#include <vector>
using namespace std;

class Solution
{
public:
    int singleNumber(vector<int> &nums)
    {
        int ones = 0;
        int twos = 0;
        for (int num : nums)
        {
            // If the bit is in twos, that means it is flipped 2 times already, hence do not account for it
            // but flip the bits to account for the num that appears for the first time, as they might the one that appears only once
            ones = (~twos) & (ones ^ num);

            // If the bit is flipped in ones, that means we have only seen 1 of that number so far, hence do not account for it
            // But flip the bits to account for the num that appears for the second time, as they are definitely not the number we want
            twos = (~ones) & (twos ^ num);
        }
        return ones;
    }
};

File: q0137/q0137.java
package q0137;

class Solution {
    public int singleNumber(int[] nums) {
        int ones = 0;
        int twos = 0;
        for (int num : nums) {
            // Register the bits appeared in the new number
            // Clear the the bits that have appeared two times
            ones = (num ^ ones) & (~twos);

            // Register the bits appeared in the new number
            // Clear the bits that have only appeared once so far
            twos = (~ones) & (num ^ twos);
        }
        return ones;
    }
}